src/memray/_memray.pyx

```diff
@@ -318,97 +321,130 @@ def start_thread_trace(frame, event, arg):
 cdef class FileReader:
     cdef cppstring _path
 
-    cdef shared_ptr[RecordReader] _reader
-    cdef vector[_Allocation] _allocations
+    cdef object _file
     cdef vector[_MemoryRecord] _memory_records
-    cdef unique_ptr[HighWatermark] _high_watermark
-    cdef bool _closed
+    cdef HighWatermark _high_watermark
     cdef object _header
 
     def __init__(self, object file_name):
-        self._path = str(file_name)
-        if not pathlib.Path(self._path).exists():
-            raise IOError(f"No such file: {self._path}")
-        self._reader = make_shared[RecordReader](unique_ptr[FileSource](new FileSource(self._path)))
-        self._header: dict = self._reader.get().getHeader()
-        self._populate_allocations()
+        try:
+            self._file = open(file_name)
+        except OSError as exc:
+            raise OSError(f"Could not open file {file_name}: {exc.strerror}") from None
 
-    cdef void _populate_allocations(self):
-        cdef RecordReader* reader = self._get_reader()
+        self._path = "/proc/self/fd/" + str(self._file.fileno())
+        self._populate_metadata()
```

Comment:
This is a bit misleading since we have metadata in the file and the method actually populates a ton of extra stuff. (I know, naming, amiright? 😄 )

Reply:
Suggestions welcome. It populates the memory records and the header, and finds the high water mark.

Or I can just inline it into the `__cinit__` and avoid giving it a name at all... any objections to that approach?

Reply:
Inlining in the `__cinit__` is good 👍